src/bc_protocol/media_packet.rs

```diff
@@ -0,0 +1,381 @@
+        // First two bytes after the header for the audio stream
+        // Since AAC and ADMPC streams start in a predicatble manner
+        if data.len() < 4 {
+            return MediaDataKind::Invalid;
+        }
+        const MAGIC_VIDEO_INFO_V1: &[u8] = &[0x31, 0x30, 0x30, 0x31];
+        const MAGIC_VIDEO_INFO_V2: &[u8] = &[0x31, 0x30, 0x30, 0x32];
+        const MAGIC_AAC: &[u8] = &[0x30, 0x35, 0x77, 0x62];
+        const MAGIC_ADPCM: &[u8] = &[0x30, 0x31, 0x77, 0x62];
+        const MAGIC_IFRAME: &[u8] = &[0x30, 0x30, 0x64, 0x63];
+        const MAGIC_PFRAME: &[u8] = &[0x30, 0x31, 0x64, 0x63];
+
+        let magic = &data[..4];
+        let kind = match magic {
+            MAGIC_VIDEO_INFO_V1 | MAGIC_VIDEO_INFO_V2 => MediaDataKind::InfoData,
+            MAGIC_AAC => MediaDataKind::AudioDataAac,
+            MAGIC_ADPCM => MediaDataKind::AudioDataAdpcm,
+            MAGIC_IFRAME => MediaDataKind::VideoDataIframe,
+            MAGIC_PFRAME => MediaDataKind::VideoDataPframe,
+            _ if data.len() == CHUNK_SIZE => MediaDataKind::Continue,
+            _ => {
+                trace!("Unknown magic kind: {:x?}", &magic);
+                MediaDataKind::Unknown
+            }
+        };
+
+        // I've never had this fail yet. It checks more bytes then just the magic
+        // Including some 2 bytes at the start of the data
+        if !MediaData::full_header_check_from_kind(kind, &data) {
+            return MediaDataKind::Invalid;
+        } else {
+            return kind;
+        }
+    }
+
+    pub fn kind(&self) -> MediaDataKind {
+        MediaData::kind_from_raw(&self.data)
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+
+    pub fn full_header_check_from_kind(kind: MediaDataKind, data: &[u8]) -> bool {
+        // This will run more advanced checks to ensure it is a valid header
+        let header_size = MediaData::header_size_from_kind(kind);
+        if data.len() < header_size {
+            trace!("Magic failed header checks on size: {:x?}", &data);
+            return false;
+        }
+
+        match kind {
+            MediaDataKind::VideoDataIframe => {
+                let stream_type = &data[4..8];
+                if let Ok(stream_type_name) = str::from_utf8(stream_type) {
+                    match stream_type_name {
+                        "H264" => true,
+                        "H265" => true,
+                        _ => {
+                            trace!(
+                                "Video Iframe failed header checks: {:x?}",
+                                &data[0..min(MAX_HEADER_SIZE, data.len())]
+                            );
+                            false
+                        }
+                    }
+                } else {
+                    false
+                }
+            }
+            MediaDataKind::VideoDataPframe => {
```

Comment:
The Iframe and Pframe cases are identical, collapse them.

Reply:
Done